bundles/org.openhab.ui.basic/src/main/java/org/openhab/ui/basic/internal/render/InputRenderer.java

```diff
@@ -58,38 +75,282 @@ public boolean canRender(Widget w) {
 
     @Override
     public EList<Widget> renderWidget(Widget w, StringBuilder sb, String sitemap) throws RenderException {
-        String snippet = getSnippet("input");
+        Input input = (Input) w;
 
+        String snippet = getSnippet("input");
         snippet = preprocessSnippet(snippet, w);
 
-        String dataState = getValue(w);
-        State state = itemUIRegistry.getState(w);
-        if (state == null || state instanceof UnDefType) {
-            snippet = snippet.replace("%undef_state%", dataState);
-            snippet = snippet.replace("%data_state%", "");
-        } else {
-            snippet = snippet.replace("%undef_state%", "");
-            snippet = snippet.replace("%data_state%", dataState);
-        }
-
         Item item = null;
         try {
             item = itemUIRegistry.getItem(w.getItem());
         } catch (ItemNotFoundException e) {
             logger.debug("Failed to retrieve item during widget rendering: {}", e.getMessage());
+            return ECollections.emptyEList();
         }
+
         String dataType;
-        if (item != null && item.getAcceptedDataTypes().stream().anyMatch(o -> Number.class.isAssignableFrom(o))) {
-            dataType = "Number";
+        if (item.getAcceptedCommandTypes().stream()
+                .anyMatch(o -> (!StringType.class.isAssignableFrom(o) && Number.class.isAssignableFrom(o)))) {
+            dataType = "number";
+        } else if (item.getAcceptedCommandTypes().stream()
+                .anyMatch(o -> (!StringType.class.isAssignableFrom(o) && DateTimeType.class.isAssignableFrom(o)))) {
+            dataType = "datetime";
         } else {
-            dataType = "Text";
+            dataType = "text";
         }
         snippet = snippet.replace("%data_type%", dataType);
 
+        String inputHint = input.getInputHint();
+        List<Class<? extends State>> dataTypes = item.getAcceptedDataTypes();
+        if (("number".equals(inputHint)
+                && !(dataTypes.contains(DecimalType.class) || dataTypes.contains(PercentType.class)))
+                || (("date".equals(inputHint) || "time".equals(inputHint) || "datetime".equals(inputHint))
+                        && !dataTypes.contains(DateTimeType.class))) {
+            logger.warn("Invalid inputHint {} for item {} of type {}, set to default", inputHint, item.getName(),
+                    item.getType());
+            inputHint = null;
+        }
+        if ("datetime".equals(dataType) && ((inputHint == null) || "".equals(inputHint) || "text".equals(inputHint))) {
```

Comment:
Would be better to use `inputHint.isEmpty()` instead of `"".equals(inputHint)`

Reply:
Adjusted and rebased.